### PR TITLE
Change default value of `action.destructive_requires_name` to True.

### LINKED ISSUE #61074: Change `action.destructive_requires_name` to true by default to avoid data deleted by accident
Forking from the ask in https://github.com/elastic/elasticsearch/issues/30878, which contains two separated issue.
Changing the default  `action.destructive_requires_name` to true seems a legitimate request and could save a few disastrous situations. 

If I read the following line correctly, it still defaults to `false` in 7.8.1?
https://github.com/elastic/elasticsearch/blob/v7.8.1/server/src/main/java/org/elasticsearch/action/support/DestructiveOperations.java#L36

### PR DESCRIPTION
**Pull request by YashJipkate**
This PR sets the default value of `action.destructive_requires_name` to True. Fixes #61074.

### CODE REVIEW

**Comment by matriv** [comment]
@YashJipkate Please check references of this setting in the docs and other places like for example: `distribution/src/config/elasticsearch.yml` and make the appropriate changes.

File: distribution/src/config/elasticsearch.yml
Comment (YashJipkate) [inline distribution/src/config/elasticsearch.yml:88]: Since the setting is now set to true by default, I think that this isn't needed anymore.

File: docs/reference/index-modules/blocks.asciidoc
Comment (YashJipkate) [inline docs/reference/index-modules/blocks.asciidoc:79]: Here and elsewhere, the doc earlier said about 'changing' the setting, so I think just an assertive statement would suffice.

**Review by YashJipkate — COMMENTED** [review]
@matriv There are some changes I made to the doc. Please let me know if you find any issues.

**Comment by matriv** [comment]
I think we should also change in `docs/reference/indices/open-close.asciidoc` L53 the phrasing since now it's by default `true. Also in `docs/reference/modules/indices/index_management.asciidoc` L16.

File: distribution/src/config/elasticsearch.yml
Comment (matriv) [reply distribution/src/config/elasticsearch.yml:88]: I'm not 100% sure, I'd rather have someone else to double check this.

**Comment by matriv** [comment]
@javanna Could you please take a look as well?

**Comment by YashJipkate** [comment]
> I think we should also change in `docs/reference/indices/open-close.asciidoc` L53 the phrasing since now it's by default `true`. Also in `docs/reference/modules/indices/index_management.asciidoc` L16.

I too thought of rephrasing them at first, but they didn't say anything about 'changing' the setting to `true`. They merely state a condition, which is now satisfied by default. 
 
Although I'm open to suggestions on how to phrase it even better.

**Comment by javanna** [comment]
jenkins test this please

**Comment by YashJipkate** [comment]
@javanna @matriv I am not able to understand the cause for the failure of so many tests. I only changed one line in actual code [here](https://github.com/elastic/elasticsearch/blob/714bb7fd26afbfe6f4e0efb225baec07e822e38b/server/src/main/java/org/elasticsearch/action/support/DestructiveOperations.java#L36). Could you please help me resolve these.

**Comment by javanna** [comment]
@YashJipkate  I quickly looked and I think that our integration tests rely on deleting all indices without specifying their names. This is done before every single test, which explains why every single integration test fails ;)

**Comment by rjernst** [comment]
Separate from the tests, I have two comments:
* Please make the PR title more precise, referring to the actual setting being change and what the value is changing to
* Since this is a change in behavior, this should only be done in a major version. We may want to give a deprecation warning for this upcoming change in behavior in 7.x. @javanna wdyt?

**Comment by YashJipkate** [comment]
> Please make the PR title more precise, referring to the actual setting being change and what the value is changing to

@rjernst Done.

**Comment by YashJipkate** [comment]
> @YashJipkate I quickly looked and I think that our integration tests rely on deleting all indices without specifying their names. This is done before every single test, which explains why every single integration test fails ;)

In that case, could you please guide me as to what can be done to resolve this?

**Comment by javanna** [comment]
@rjernst I was wondering the same about this change in behaviour. I agree it should be done in a major. I don't think I will get to properly review this PR, would you mind finding somebody to help moving this forward please?

File: distribution/src/config/elasticsearch.yml
Comment (williamrandolph) [reply distribution/src/config/elasticsearch.yml:88]: I recommend leaving this block, but changing the description to "allow wildcard deletion of indices" and changing the value from "true" to "false."

File: docs/reference/index-modules/blocks.asciidoc
Comment (williamrandolph) [reply docs/reference/index-modules/blocks.asciidoc:79]: I think it's a little bit confusing that the first sentance says to use wildcards, and then the second sentance proceeds to explain that this is disallowed. I'd recommend rewording the whole paragraph as follows:

> By default, you must explicitly name the indices you are adding blocks to. To allow the adding of blocks to indices with `_all`, `*`, or other wildcard expressoins, change the `action.destructive_requires_name setting to `false`. You can update this setting...

File: docs/reference/indices/close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/close.asciidoc:37]: As above, I think it's worth rewording this paragraph so that the default behavior is explained before the configuration option.

The same recommendation applies to the other docs files.

**Review by williamrandolph — CHANGES_REQUESTED** [review]
In order to fix the tests, we now need to set `action.destructive_requires_name` cluster to false for our test clusters. It looks to me like the easiest place to do that is in the code for the test cluster setup, in the file `buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java`. Specifically, I added:

```
baseConfig.put("action.destructive_requires_name", "false");
```

...after line 1187 of that file, and that let some of the failing integration tests pass.

I'm not sure if this will fix everything, so we may have to iterate a little bit to allow wildcard deletions in every place where the tests expect them.

I requested some wording changes in other comments and meant for those requests to be part of this review.

Thank you for working with us on this PR!

File: distribution/src/config/elasticsearch.yml
Comment (YashJipkate) [reply distribution/src/config/elasticsearch.yml:88]: Done.

File: docs/reference/index-modules/blocks.asciidoc
Comment (YashJipkate) [reply docs/reference/index-modules/blocks.asciidoc]: Done.

File: docs/reference/indices/close.asciidoc
Comment (YashJipkate) [reply docs/reference/indices/close.asciidoc]: Done.

**Comment by YashJipkate** [comment]
@williamrandolph I've done the changes requested by you.

Also, can you start the Jenkins here since the tests are being flaky on my PC (also takes too long and freezes the system, times out), as it isn't running automatically for the subsequent pushes on this PR. Or is there any other way I can carry out the tests?

**Comment by williamrandolph** [comment]
@elasticmachine ok to test

File: docs/reference/indices/close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/close.asciidoc:38]:
```suggestion
By default, you must explicitly name the indices you are closing.
To specify indices to close with `_all`, `*``, or other wildcard
expressions, change the `action.destructive_requires_name` setting to `false`.
```

File: docs/reference/indices/delete-index.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/delete-index.asciidoc:37]:
```suggestion
By default, you must explicitly name the indices you are deleting.
To specify indices to delete with `_all`, `*``, or other wildcard
expressions, change the `action.destructive_requires_name` setting to `false`.
You can update this setting in the `elasticsearch.yml` file or using the
```

File: docs/reference/indices/open-close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/open-close.asciidoc:51]:
```suggestion

```

File: docs/reference/indices/open-close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/open-close.asciidoc:55]:
```suggestion
By default, you must explicitly name the indices you are opening or closing.
To open or close indices with `_all`, `*``, or other wildcard
expressions, change the `action.destructive_requires_name` setting to `false`.
```

File: docs/reference/indices/open-close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/open-close.asciidoc:91]:
```suggestion

```

File: docs/reference/indices/open-close.asciidoc
Comment (williamrandolph) [inline docs/reference/indices/open-close.asciidoc:95]:
```suggestion
By default, you must explicitly name the indices you using to limit the request.
To limit a request using `_all`, `*``, or other wildcard
expressions, change the `action.destructive_requires_name` setting to `false`.
```

**Review by williamrandolph — CHANGES_REQUESTED** [review]
It looks like you copied and pasted a documentation update without changing its context. I've added a suggestion for each section. Feel free to use your own phrasing, but make sure that each update matches the context of its section.

I believe I've enabled tests on this PR for future commits, but if you don't see tests running please comment and I will make sure they run.

**Comment by williamrandolph** [comment]
I'll look at these test failures and see if I can point to targeted fixes for their configuration.

**Comment by YashJipkate** [comment]
> It looks like you copied and pasted a documentation update without changing its context.

The previous statements looked the same so I wrongly judged that they might have the same context. Thanks for pointing that out.


> I'll look at these test failures and see if I can point to targeted fixes for their configuration.

Thank you lots 😄, I'll also try to fix some of them.

**Comment by williamrandolph** [comment]
@YashJipkate Some of the tests can be fixed by updating the `newNode()` method in the `ESSingleNodeTestCase` class. Specifically, we need to add `.put(DestructiveOperations.REQUIRES_NAME_SETTING.getKey(), false)` to the code that constructs the setting list in that method.

**Comment by YashJipkate** [comment]
@williamrandolph I made the changes you requested. 5 checks are now passing.

**Comment by williamrandolph** [comment]
@YashJipkate A few more notes on failing tests:

The `DateMathExpressionResolverTests` failure in `elasticsearch-ci/1` doesn't have anything to do with your work. It's been fixed on the Elasticsearch master branch.

Some or all of the failing tests in `elasticsearch-ci/2` can be fixed by updating the settings list in `test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java`, by putting `builder.put(DestructiveOperations.REQUIRES_NAME_SETTING.getKey(), false);` just above line 402.

I'm still looking at the other test failures.

**Comment by williamrandolph** [comment]
@YashJipkate I think some of the test failures will disappear if we merge the latest master branch into this PR. Let me know if you'd like me to update this branch with latest master. It's also fine if you can do it.

Thanks again for your patience with these test class changes. There are many different kinds of tests in this repository and it can be frustrating to track down fixes for all of them.

**Comment by williamrandolph** [comment]
@elasticmachine update branch

@YashJipkate I'm going to go ahead and update the branch so we can see what's fixed by merging master.

**Comment by williamrandolph** [comment]
@YashJipkate 

I think we can fix almost all of the remaining failures with changes in four files:

**test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java** - defines some settings in the `nodeSettings()` method, and we need to allow wildcard deletions here

**qa/remote-clusters/docker-compose-oss.yml** - this file defines two different docker containers, both of them need their settings changed to allow wildcard deletions
**qa/remote-clusters/docker-compose.yml** - this file _also_ defines two different docker containers, both of them need their settings changed to allow wildcard deletions

**qa/os/src/test/java/org/elasticsearch/packaging/util/ServerUtils.java** - in this file, the `runElasticsearchTests()` method cleans up after itself by calling DELETE `_all`. We can delete the `library` index here by name instead.

**Comment by YashJipkate** [comment]
@williamrandolph Apologies for a late update, having some health issues currently. I have done the changes requested by you.

> @YashJipkate I'm going to go ahead and update the branch so we can see what's fixed by merging master.

Thank you!

**Comment by YashJipkate** [comment]
@williamrandolph all but two are now passing.

`elasticsearch-ci/1` is failing because it also needs to update the value in some setting. But I can't guess which one. I tried updating the `restClientSettings` but it doesn't seem to work.

`elasticsearch-ci/default-distro` is failing because I didn't 'import' `action.support.DestructiveOperations`. I am not able to figure out how to do that.

File: qa/remote-clusters/docker-compose.yml
Comment (williamrandolph) [inline qa/remote-clusters/docker-compose.yml:33]: This should be `action.destructive_requires_name=false`, here and in other `docker-compose.yml` files. Making this change should take care of the unknown setting failure message.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java
Comment (williamrandolph) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:1748]: Would you mind moving this setting either above the comment or below the `LOW_LEVEL_CANCELLATION_SETTING` line? Right now it's between a comment and the settings that the comment refers to, which isn't ideal.

**Review by williamrandolph — CHANGES_REQUESTED** [review]
It looks like the failures in elasticsearch-ci/1 are because we need to add settings to all four of the instances defined in `distribution/docker/docker-compose.yml`.

I think we are getting close to approval now.

File: qa/remote-clusters/docker-compose.yml
Comment (YashJipkate) [reply qa/remote-clusters/docker-compose.yml]: Done.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java
Comment (YashJipkate) [reply test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java]: Moved up.

**Comment by williamrandolph** [comment]
@YashJipkate I got a notification on some failing tests that we hadn't seen before. Merging master into your branch will fix the `elasticsearch-ci/bwc` failures, at the very least.

**Comment by YashJipkate** [comment]
@williamrandolph The tests are finally passing!

**Review by williamrandolph — APPROVED** [review]
@YashJipkate Thanks for iterating on this with us. You probably didn't expect so many test failures from such a small change, but we appreciate your seeing it through.

I'm going to approve this pull request, but since this is a breaking change I need to hold off on merging it for a week or so in order to give other teams at Elastic time to prepare for the change. I hope you understand about the delay.

**Comment by YashJipkate** [comment]
> @YashJipkate Thanks for iterating on this with us. You probably didn't expect so many test failures from such a small change, but we appreciate your seeing it through.

But I feel it was worth it, I had a very good experience :)


> I'm going to approve this pull request, but since this is a breaking change I need to hold off on merging it for a week or so in order to give other teams at Elastic time to prepare for the change. I hope you understand about the delay.

Yeah, absolutely.

**Comment by spalger** [comment]
I'm trying to figure out what changes will be necessary in Kibana to make this work in https://github.com/elastic/kibana/pull/88986 and the first thing I'm noticing is that I'm getting a wildcard error when trying to delete an index with a concrete index name, but when the index doesn't exist:

```
Elasticsearch TRACE: 2021-01-21T18:48:30Z
  -> DELETE https://localhost:9220/functional-test-actions-index?ignore_unavailable=true

  <- 400
  {
    "error": {
      "root_cause": [
        {
          "type": "illegal_argument_exception",
          "reason": "Wildcard expressions or all indices are not allowed"
        }
      ],
      "type": "illegal_argument_exception",
      "reason": "Wildcard expressions or all indices are not allowed"
    },
    "status": 400
  }
```

This feels like a bug, maybe even one that's unrelated to this PR, but figured I'd start by bringing it up here and see what y'all think we should do about this.

**Comment by williamrandolph** [comment]
@spalger I just did a quick check with 7.10.2 and I don't see any similar behavior there with either setting of `action.default_requires_name`, so I'll keep looking into this issue and why it comes up on this branch. Thanks for reporting it.

**Comment by williamrandolph** [comment]
@spalger I have tried a couple of versions of trying to reproduce this with and without other indices or hidden indices and can't get the warning in the way you reported it. Is there anything else you can provide to help me reproduce it?
```
# create index
➜  elasticsearch-7.10.2 curl -s -XPUT "localhost:9200/functional-test-actions-index"
{
  "acknowledged":true,
  "shards_acknowledged":true, 
  "index":"functional-test-actions-index"
}

# index is there to delete                           
➜  elasticsearch-7.10.2 curl -XDELETE 'localhost:9200/functional-test-actions-index?ignore_unavailable=true' | jq '.'
{
  "acknowledged": true
}

# index is not there, but ignore-unavailable works
➜  elasticsearch-7.10.2 curl -s -XDELETE 'localhost:9200/functional-test-actions-index?ignore_unavailable=true' | jq '.'
{
  "acknowledged": true
}

# using a wildcard - destructive_requires_name is set correctly
➜  elasticsearch-7.10.2 curl -s -XDELETE 'localhost:9200/functional-*?ignore_unavailable=true' | jq '.'
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Wildcard expressions or all indices are not allowed"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Wildcard expressions or all indices are not allowed"
  },
  "status": 400
}
```

**Comment by spalger** [comment]
I'm running a master snapshot built by Kibana's ES Snapshot mechanism, which we use to allow tracking master without breaking CI when breaking changes are merged into ES. The snapshot can be downloaded from https://storage.googleapis.com/kibana-ci-es-snapshots-daily/8.0.0/archives/20210124-194645_63c85ff3e84/elasticsearch-8.0.0-SNAPSHOT-darwin-x86_64.tar.gz

We are currently running ES with security enabled, which maybe has something to do with it, and the CLI options: `-E indices.query.bool.max_nested_depth=100 -E action.destructive_requires_name=true`

I can reproduce the issue with a single request:
```
curl -s -XDELETE 'http://elastic:changeme@localhost:9200/functional-test-actions-index?ignore_unavailable=true&pretty'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "Wildcard expressions or all indices are not allowed"
      }
    ],
    "type" : "illegal_argument_exception",
    "reason" : "Wildcard expressions or all indices are not allowed"
  },
  "status" : 400
}
```

**Comment by williamrandolph** [comment]
@spalger Thanks for the additional information and clean reproduction line! It looks like having security enabled makes the difference. I've reproduced the issue on the 7.10.2 release, so it's not specific to this PR and I won't consider it a blocker for merging. I will, however, make sure we have an issue covering the bug and link it back to this PR.

**Comment by spalger** [comment]
On second thought I don't think it's a good idea to update Kibana in a bunch of places to handle ES responding with 400 instead of 404. If you decide to merge this PR before https://github.com/elastic/elasticsearch/issues/67958 is fixed then Kibana will only work with `action.destructive_requires_name=false`, which we can use by default for most Kibana dev and testing, but it means Kibana master breaks in pretty major ways when run against ES master which is a state we don't want to be in for very long. Basically, I don't think I'm in any way eligible to dictate priorities here but it would make things a good deal more stable for us if we could fix https://github.com/elastic/elasticsearch/issues/67958 first.

**Comment by williamrandolph** [comment]
@spalger Thanks for the feedback. I've got PR https://github.com/elastic/elasticsearch/pull/68021 in progress to fix the issue with unexpected wildcard warnings and Elasticsearch security, and I'll try to get it in ASAP so that you can verify the fix before this one is merged.

**Comment by williamrandolph** [comment]
@elasticmachine update branch

**Comment by spalger** [comment]
I think that makes us g2g, much appreciated!

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Pull request merged**
